### PR TITLE
Implement 802.1ah I-Tag for PBB MAC-in-MAC encapsulation

### PR DESCRIPTION
**Pull request by alexforencich**
Implement IEEE 802.1ah I-Tag for Provider Backbone Bridging (PBB) MAC-in-MAC encapsulation.

The tag is similar to a VLAN tag, but it is effectively followed by a complete Ethernet packet including addresses, instead of only the Ethertype and payload.  

Also, there are probably at least 3 different ways of supporting 802.1ah, I have implemented one way but I'm not sure it's necessarily the best way.  Wireshark at least seems to subsume the B-tag (802.1ad type 0x88a8 outer/Q-in-Q VLAN tag), I-tag (this tag, type 0x88e7), and the following Ethernet header (as C-Destination, C-Source, and Type fields) all under the same IEEE 802.1ah block.  I'm not sure if it makes sense to do something similar in scapy, or if the tags should be kept separate.  The IEEE spec seems to consider the "wrapped" MAC addresses as part of the I-Tag, as the tag sits in between the MAC addresses and ethertype field of the original frame.

Options:

1. Everything separate (Dot1AD, Dot1AH, Ether) - what I have implemented here so far
2. Pull Ethernet header fields of wrapped frame into Dot1AH
3. Pull Ethernet headers and preceding Dot1AD into Dot1AH (what Wireshark seems to do)

I'm also not sure if I have done everything correctly wrt. scapy internals, particularly in relation to aliastypes, register_l3, and default_payload_class.  It seems to build and dissect correctly, but I have not done much testing yet.  I also don't have a unit test implemented - which file should I add it to?

### CODE REVIEW

**Comment by alexforencich** [comment]
Here is a python script that generates a handful of different encapsulated frames including some PBB frames, along with the output pcap file, for examination in wireshark.

[encap_test.zip](https://github.com/secdev/scapy/files/11468894/encap_test.zip)

**Comment by alexforencich** [comment]
Interesting, from the test failures it looks like 802.1ah is already implemented in scapy as "SPBM" under contrib.  Presumably that should be renamed to Dot1AH and moved into l2.py with the contrib implementation replaced with an alias for backwards compatibility, as the 802.1aq standard (SPB) references the tag format defined in 802.1ah (PBB).

**Comment by alexforencich** [comment]
Updated the PR to move the `SPBM` implementation to `Dot1AH` under `scapy.layers.l2`, replacing `SPBM` with an alias to `Dot1AH` for backwards-compatibility.

**Comment by alexforencich** [comment]
Added unit tests to l2.uts for `Dot1Q`, `Dot1AD`, and `Dot1AH`.  I think this is probably good enough to take out of draft mode, although there still may need to be some adjustments for `aliastypes`, `register_l3`, and `default_payload_class`.

File: scapy/layers/l2.py [hunk 1]
```diff
@@ -238,6 +239,7 @@ def i2m(self, pkt, x):
 }
 
 ETHER_TYPES[0x88a8] = '802_AD'
+ETHER_TYPES[0x88e7] = '802_AH'
```
[the remark this answers is not in the document]
Comment (alexforencich) [reply scapy/layers/l2.py:242]: Pattern extrapolation from the previous one.  Probably they should both be switched, but I also don't know scapy well enough to know exactly what these do.

**Comment by alexforencich** [comment]
> Personally I think the old name should be dropped, to avoid propagating confusion any further. Do we actually need to maintain that compatibility here?

Well, it's in contrib, so I'm not sure what makes the most sense.  Removing it could break someone's use of scapy, presumably.  But the alias is literally just `SPBM = Dot1AH`.  Could maybe leave it in for a release or two, then remove the alias.  Possibly also mark as deprecated somehow.  Also, there is a unit test for `SPBM` that I didn't touch.

**Comment by alexforencich** [comment]
> > > Personally I think the old name should be dropped, to avoid propagating confusion any further. Do we actually need to maintain that compatibility here?
> > 
> > 
> > Well, it's in contrib, so I'm not sure what makes the most sense. Removing it could break someone's use of scapy, presumably. But the alias is literally just `SPBM = Dot1AH`. Could maybe leave it in for a release or two, then remove the alias. Possibly also mark as deprecated somehow. Also, there is a unit test for `SPBM` that I didn't touch.
> 
> What if someone actually decides to implement handling of IEEE 802.1aq / SPB in Scapy though? Don't we want to reserve the name `SPBM` for that? (That said, I believe the only PDUs added by SPB are really just additional TLVs for existing PDUs in IS-IS, so perhaps there's no concern here.)

Well, if you think it would be best to delete it completely, then I can certainly do that instead replacing it with an alias.

**Comment by alexforencich** [comment]
> > 5\. Pull Ethernet headers and preceding Dot1AD into Dot1AH (what Wireshark seems to do)
> 
> I-tagged service interfaces (section 25.5 of the specification) do not include a B-TAG. It looks like Wireshark includes the B-TAG as part of the 802.1ah dissector if it is there, but ignores it if it is not.
> 
> So, if you want to match specifically on the B-VID (and not the S-VID), or vice versa, I believe this was the only way to do it before Wireshark 4.0 - so I assume that's why they did it. But now, you can filter on a specific occurrence of the same header in Wireshark (with the `#` notation). Scapy also lets you do this, with the `nb` parameter to `scapy.packet.getlayer()`.

Right, I guess that makes sense.  So implementing it the way I have done here (and how `SPBM` was implemented) is probably the best method.

**Comment by alexforencich** [comment]
One other thing to sort out: `Dot1Q` has a bunch of additional stuff in it (aliastypes, answers, default_payload_class, extractpadding, and mysummary).  How much of that should be applied to `Dot1AH`?

**Review by gpotter2 — APPROVED** [review]
Hi & sorry for the delay.
This looks good, thanks for keeping the compatibility !

**Comment by alexforencich** [comment]
If you want to delete spbm.py, you'll also need to delete or update the existing spbm unit test.  

What about the extra stuff in `Dot1Q` (aliastypes, answers, default_payload_class, extractpadding, and mysummary) ?  Does any of that need to be transferred over to `Dot1AH`?

**Comment by gpotter2** [comment]
> What about the extra stuff in Dot1Q (aliastypes, answers, default_payload_class, extractpadding, and mysummary) 

`answers` and `mysummary` sure. The others I don't think a necessary.

**Comment by alexforencich** [comment]
Got it, makes sense.  I suppose `ETHER_TYPES[0x88e7] = '802_AH'` should probably also be adjusted to `802_1AH`, and possibly same thing with the existing `802_AD` which @dpward noticed earlier.  Is there anything else that needs to be done with this so it's ready for merging?

File: scapy/layers/l2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexforencich) [reply scapy/layers/l2.py]: Looks like @gpotter2 added a commit to fix these

**Review by gpotter2 — APPROVED** [review]

**Pull request merged**